src/metabase/lib/equality.cljc

```diff
@@ -145,3 +134,30 @@
          [:field opts (field-id :guard integer?)]
          (when-let [field-name (:name (lib.metadata/field metadata-providerable field-id))]
            (find-closest-matching-ref [:field opts field-name] refs))))))
+
+(defn mark-selected-columns
+  "Mark `columns` as `:selected?` if they appear in `selected-columns-or-refs`. Uses fuzzy matching
+  with [[find-closest-matching-ref]].
+
+  Example usage:
+
+    ;; example (simplified) implementation of [[metabase.lib.field/fieldable-columns]]
+    ;;
+    ;; return (visibile-columns query), but if any of those appear in `:fields`, mark then `:selected?`
+    (mark-selected-columns (visibile-columns query) (:fields stage))"
+  ([columns selected-columns-or-refs]
+   (mark-selected-columns nil columns selected-columns-or-refs))
+
+  ([metadata-providerable columns selected-columns-or-refs]
+   (let [selected-refs          (mapv lib.ref/ref selected-columns-or-refs)
+         refs                   (mapv lib.ref/ref columns)
+         matching-selected-refs (into #{}
+                                      (map (fn [selected-ref]
+                                             (if metadata-providerable
+                                               (find-closest-matching-ref metadata-providerable selected-ref refs)
+                                               (find-closest-matching-ref selected-ref refs))))
```

Comment:
Could `find-closest-matching-ref` be made smart about a `nil` `metadata-providerable` instead?

Reply:
good call. Fixed